### PR TITLE
Added compatibility for IEEE 754 half-precision floats 

### PR DESCRIPTION
**Pull request by vetsE**
Small modification of the core.
I mostly mirrored what's was already in place for Float32's for Float16's.

### CODE REVIEW

**Comment by arekbulski** [comment]
Interesting. Could you show me some official documentation regarding (1) the e format specifier and (2) how is it officially called, a semi?

**Comment by vetsE** [comment]
(1) It's an addition to [the Python 3.6 version of struct](https://docs.python.org/3.6/library/struct.html#format-characters).
(2) Well, the official IEEE documentation calls it "binary16". I thought "semi" or "half" would be more in line with "single" and "double". I used "semi" mostly because that's what the ZigBee Cluster Library calls that type. I do admit, after investigating, that "half" is more frequent and would make more sense.

**Comment by arekbulski** [comment]
Half is fine.

If you look at Travis, the code fails on anything before Python 3.6. You would need to update both the code (using py3compat PY) and tests (using XFAIL) and docs (an annotation regarding 3.6).

**Comment by vetsE** [comment]
My bad.

I updated my branch. The code doesn't fail on Travis anymore. And I did annotate the docs.

File: construct/core.py [hunk 1]
```diff
@@ -1376,6 +1394,8 @@ def Float64n():
     """Native endianity, 64-bit IEEE floating point number"""
     return FormatField("=", "d")
 
+if supportshalffloats:
```
Comment (arekbulski) [inline construct/core.py]: Remove IF.

File: construct/core.py [hunk 2]
```diff
@@ -1350,6 +1354,20 @@ def Int64sn():
 Int   = Int32ub
 Long  = Int64ub
 
+if supportshalffloats:
```
Comment (arekbulski) [inline construct/core.py]: Remove IF.

File: construct/core.py [hunk 3]
```diff
@@ -1017,8 +1017,12 @@ class FormatField(Construct):
     def __init__(self, endianity, format):
         if endianity not in list("=<>"):
             raise FormatFieldError("endianity must be like: = < >", endianity)
-        if format not in list("fdBHLQbhlq"):
-            raise FormatFieldError("format must be like: f d B H L Q b h l q", format)
+        supported_formats = "fdBHLQbhlq"
```
Comment (arekbulski) [inline construct/core.py]: Remove IF.

File: construct/__init__.py [hunk 4]
```diff
@@ -207,3 +207,8 @@
 ]
 __all__ += ["Int%s%s%s" % (n,us,bln) for n in (8,16,24,32,64) for us in "us" for bln in "bln"]
 __all__ += ["Float%s%s" % (n,bln) for n in (32,64) for bln in "bln"]
+
+if lib.supportshalffloats:
```
Comment (arekbulski) [inline construct/__init__.py]: Remove IF.

File: construct/core.py [hunk 5]
```diff
@@ -1017,8 +1017,9 @@ class FormatField(Construct):
...
         if endianity not in list("=<>"):
             raise FormatFieldError("endianity must be like: = < >", endianity)
-        if format not in list("fdBHLQbhlq"):
-            raise FormatFieldError("format must be like: f d B H L Q b h l q", format)
+
+        if format not in list("fdBHLQbhlqe"):
+            raise FormatFieldError("format must be like: f d B H L Q b h l q e", format)
         super(FormatField, self).__init__()
         self.fmtstr = endianity+format
         self.length = struct.calcsize(endianity+format)
```
Comment (arekbulski) [inline construct/core.py]: This line calculates the size for "e" regardless of Python version, thus exception. You could just put an IF on this line, that would solve the problem.

**Comment by vetsE** [comment]
Ah, I see. Makes sense.

I amended my commit with what you suggested.

**Comment by RenaKunisaki** [comment]
There's also a variant used by ARM, which is identical except that there are no Inf/NaN values; the values that would encode those are instead treated normally, allowing a larger range of numbers to be represented. Any chance to support this?

**Pull request merged**